File: 912-sort-array/CountingSort.java
import java.util.Arrays;
public class CountingSort {
    public static int[] sort(int[] nums){
        if (nums == null)
            return null;
        int max = nums[0];
        int min = nums[0];
        for (int val :nums){
            if (val > max)
                max = val;
            else if (val < min)
                min = val;
        }
        int trueRange = max - min + 1;
        if (trueRange == 0)
            return null;

        int[] counting = new int[trueRange];

        for (int val : nums) {
            counting[val - min] += 1;
        }

        int preSum = 0;
        for (int index = 0; index < counting.length; index++){
            int temp = counting[index];
            counting[index] = preSum;
            preSum += temp;
        }
        System.out.println(Arrays.toString(counting));

        int[] result = new int[nums.length];
        for (int index = 0; index < nums.length; index++){
            result[counting[nums[index] - min]] = nums[index];
            counting[nums[index] - min]++;
        }
        System.arraycopy(result, 0, nums, 0, result.length);
        return nums;
    }
    static void main(String[] args) {
        int[] nums = {2, 4, 5, 2, 6, 4};
        sort(nums);
        System.out.println(Arrays.toString(nums));
    }
}
/*
[2, 0, 2, 1, 1]
[0, 2, 2, 4, 5]
 */

File: 912-sort-array/countingsort.go
/*
參考代碼:
algo/CountingSort/countingsort_advance.go
*/
package main

import "fmt"

func countingsort(nums []int) []int {
	minNum := arrMinInt(nums)
	maxNum := arrMaxInt(nums)
	trueRange := maxNum - minNum + 1

	counting := make([]int, trueRange)
	for _, value := range nums {
		counting[value-minNum]++
	}

	preSum := 0
	for index := 0; index < len(counting); index++ {
		temp := counting[index]
		counting[index] = preSum
		preSum += temp
	}

	result := make([]int, len(nums))
	for _, value := range nums {
		result[counting[value-minNum]] = value
		counting[value-minNum]++
	}
	return result
}
func arrMaxInt(arr []int) (maxNum int) {
	maxNum = arr[0]
	for _, value := range arr {
		maxNum = max(maxNum, value)
	}
	return
}
func arrMinInt(arr []int) (minNum int) {
	minNum = arr[0]
	for _, value := range arr {
		minNum = min(minNum, value)
	}
	return
}

func main() {
	arr := []int{5, 7, 3, 1, 6, 8, 9, 4, 7}
	fmt.Println("Result = ", countingsort(arr))
}
